pkg/cri/sbserver/container_remove.go

```diff
@@ -33,19 +33,26 @@ import (
 // RemoveContainer removes the container.
 func (c *criService) RemoveContainer(ctx context.Context, r *runtime.RemoveContainerRequest) (_ *runtime.RemoveContainerResponse, retErr error) {
 	start := time.Now()
-	container, err := c.containerStore.Get(r.GetContainerId())
+	ctrID := r.GetContainerId()
+	container, err := c.containerStore.Get(ctrID)
 	if err != nil {
 		if !errdefs.IsNotFound(err) {
-			return nil, fmt.Errorf("an error occurred when try to find container %q: %w", r.GetContainerId(), err)
+			return nil, fmt.Errorf("an error occurred when try to find container %q: %w", ctrID, err)
 		}
 		// Do not return error if container metadata doesn't exist.
-		log.G(ctx).Tracef("RemoveContainer called for container %q that does not exist", r.GetContainerId())
+		log.G(ctx).Tracef("RemoveContainer called for container %q that does not exist", ctrID)
 		return &runtime.RemoveContainerResponse{}, nil
 	}
 	id := container.ID
 	i, err := container.Container.Info(ctx)
 	if err != nil {
-		return nil, fmt.Errorf("get container info: %w", err)
+		// Since containerd doesn't see the container and criservice's content store does,
+		// we should try to recover from this state by removing entry for this container
+		// from the container store as well.
+		log.G(ctx).WithError(err).Debugf("get container info failed")
+		c.containerStore.Delete(ctrID)
+		c.containerNameIndex.ReleaseByKey(ctrID)
+		return nil, nil
```

Comment:
My memory on k8s is hazy, but I remember it trying to remove containers even if one of the attempts errored out. Is there some behavior that explodes if we return an err instead of nil? If not we should probably keep returning the error, but keep the removal that you added that gets us back to truth state wise

Reply:
If we return an error, the kubelet keeps retrying but we were ending up in the same mismatched state and returning error. Are you suggesting that we cleanup what container store has when we are in this error state up return error so that kubelet retries again anyway and this time it would return nil from line 43-44 thereby verifying that the state is clean across containerd and container store?

Reply:
Oh I think I get the rationale here, we want RemoveContainer from k8s perspective to be a success after this change?

Reply:
Yes that was my intention :)

Reply:
Might want to return &RemoveContainerResponse{} instead of nil, I'm not sure how the grpc lib would handle marshaling nil